tower-http/src/follow_redirect/policy/mod.rs

```diff
@@ -0,0 +1,279 @@
+pub trait Policy<B, E> {
+    /// Invoked when the service received a response with a redirection status code (`3xx`).
+    ///
+    /// This method returns an [`Action`] which indicates whether the service should follow
+    /// the redirection.
+    fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E>;
+
+    /// Invoked right before the service makes a request, regardless of whether it is redirected
+    /// or not.
+    fn on_request(&mut self, _request: &mut Request<B>) {}
+
+    /// Try to clone a request body before the service makes a redirected request.
+    ///
+    /// If the request body cannot be cloned, return `None`.
+    ///
+    /// This is not invoked when [`B::size_hint`][http_body::Body::size_hint] returns zero,
+    /// in which case `B::default()` will be used to create a new request body.
+    ///
+    /// The default implementation returns `None`.
+    fn clone_body(&self, _body: &B) -> Option<B> {
+        None
+    }
+}
+
+impl<B, E, P> Policy<B, E> for &mut P
+where
+    P: Policy<B, E> + ?Sized,
+{
+    fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E> {
+        (**self).redirect(attempt)
+    }
+
+    fn on_request(&mut self, request: &mut Request<B>) {
+        (**self).on_request(request)
+    }
+
+    fn clone_body(&self, body: &B) -> Option<B> {
+        (**self).clone_body(body)
+    }
+}
+
+impl<B, E, P> Policy<B, E> for Box<P>
+where
+    P: Policy<B, E> + ?Sized,
+{
+    fn redirect(&mut self, attempt: &Attempt<'_>) -> Action<E> {
+        (**self).redirect(attempt)
+    }
+
+    fn on_request(&mut self, request: &mut Request<B>) {
+        (**self).on_request(request)
+    }
+
+    fn clone_body(&self, body: &B) -> Option<B> {
+        (**self).clone_body(body)
+    }
+}
+
+/// A redirection [`Policy`] with a reasonable set of standard behavior.
+///
+/// This policy limits the number of successive redirections ([`Limited`])
+/// and removes credentials from requests in cross-origin redirections ([`FilterCredentials`]).
+pub type Standard = Join<Limited, FilterCredentials>;
+
+/// A type that holds information on a redirection attempt.
+pub struct Attempt<'a> {
+    pub(crate) status: StatusCode,
+    pub(crate) location: &'a Uri,
+    pub(crate) previous: &'a Uri,
+}
+
+impl<'a> Attempt<'a> {
+    /// Returns the redirection response.
+    pub fn status(&self) -> StatusCode {
+        self.status
+    }
+
+    /// Returns the destination URI of the redirection.
+    pub fn location(&self) -> &'a Uri {
+        self.location
+    }
+
+    /// Returns the URI of the original request.
+    pub fn previous(&self) -> &'a Uri {
+        self.previous
+    }
+}
+
+/// A value returned by [`Policy::redirect`] which indicates the action
+/// [`FollowRedirect`][super::FollowRedirect] should take for a redirection response.
+#[derive(Clone)]
+pub struct Action<E> {
+    pub(crate) kind: ActionKind<E>,
+}
+
+#[derive(Clone)]
+pub(crate) enum ActionKind<E> {
+    Follow,
+    Stop,
+    Error(E),
+}
```

Comment:
I not sure, but wondering if the error case is necessary to have on `Action` or if using something like `Result<Action, E>` would work just as well.

If we do this it might also be worth making the `ActionKind` enum into `Action` and expose that:

```rust
pub enum Action {
    Follow,
    Stop,
}
```

We could still have `is_follow` and `is_stop` convenience methods.

It would make adding variants a breaking change but not sure what we might want to add. Either you follow a redirect or you don't 🤷

Reply:
Just for the record, I have considered of encoding more information in `Action` or `Attempt`, like whether `FollowRedirect` should remove credentials. But now that `FilterCredentials` exists, there is no need to hide the `ActionKind` variants, indeed.

We could make `Action` `#[non_exhaustive]`, but I neither think that we would benefit from it.